Remove the glossary tooltip inside the HTML tags

amieiro opened the pull request:
<!-- 
Thanks for contributing to GlotPress! Please, follow the GlotPress Contributing Guidelines:
https://github.com/GlotPress/GlotPress/blob/develop/CONTRIBUTING.md 
-->

## Problem

When GlotPress has an HTML tag in the glossary, it shows the glossary tooltip in the HTML tag.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/b25054f5-402c-41d6-908d-409e8dc85458)

Fix https://github.com/GlotPress/GlotPress/issues/1382 and https://github.com/GlotPress/GlotPress/issues/1385.

<!--
Please, describe what is the status/problem before the PR.
-->

## Solution

This PR excludes the glossary tooltips inside the HTML tags, splitting the original strings (singular and plural) to avoid adding the glossary tooltips to the HTML tags.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/c719a4fd-dbb8-4f80-b430-c76ae316315b)

<!--
Please, describe how this PR improves the situation.
-->

## Testing Instructions

This PR adds 2 tests:
- test_map_glossary_entries_to_translation_originals_excluding_terms_in_html_tags.
- test_map_glossary_entries_to_translation_originals_excluding_terms_in_html_tags_in_the_plural_origin.

To test it manually, you can:
1. Add an HTML tag (e.g., **strong**) to the glossary.
2. Add 2 new originals to be translated (singular and plural). E.g.:
  - Singular: `This is s strong test <strong>strong</strong>. This is another<dd>strong</dd>, very strong test with<img src="strong.img" /> images <hr/>.`
  - Plural: `Plural. This is s strong test <strong>strong</strong>. This is another<dd>strong</dd>, very strong test with<img src="strong.img" /> images <hr/>.`
4. Check that the tooltip for the `strong`word only appears outside the HTML tags.

<!-- 
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
Please, include step-by-step instructions on how to test this PR:
1. Open a original string in a project.
5. Add the translation.
6. etc. 
-->

amieiro committed:
Remove the tooltips from the HTML tags in the singular originals

amieiro committed:
Remove the tooltips from the HTML tags in the plural original

amieiro committed:
Remove an unused method

amieiro committed:
Add an extra line before the method return

amieiro committed:
Merge branch 'develop' into tooltip-html-tag

akirk commented:
Could we avoid adding multiple preg_splits by just checking the previous chunk like this?
```php
foreach ( $singular_split as $k => $chunk ) {
	// [...]
	if ( $k > 0 && ( '<' === substr( $singular_split[ $k - 1 ], -1 ) || '</' === substr( $singular_split[ $k - 1 ], -2 ) ) ) {
		$singular_combined .= $escaped_chunk;
		continue;
	}

```

akirk commented:
This patch also makes the test pass:

```diff
diff --git i/gp-templates/helper-functions.php w/gp-templates/helper-functions.php
index 2b43b50b..645a6295 100644
--- i/gp-templates/helper-functions.php
+++ w/gp-templates/helper-functions.php
@@ -230,10 +230,15 @@ function map_glossary_entries_to_translation_originals( $translation, $glossary
 	if ( is_array( $singular_split ) ) {
 		$singular_combined = '';
 
-		foreach ( $singular_split as $chunk ) {
+		foreach ( $singular_split as $k => $chunk ) {
 			// Create an escaped version for use later on.
 			$escaped_chunk = esc_translation( $chunk );
 
+			if ( $k > 0 && ( '<' === substr( $singular_split[ $k - 1 ], -1 ) || '"' === substr( $singular_split[ $k - 1 ], -1 ) || '</' === substr( $singular_split[ $k - 1 ], -2 ) ) ) {
+				$singular_combined .= $escaped_chunk;
+				continue;
+			}
+
 			// Create a lower case version to compare with the glossary terms.
 			$lower_chunk = strtolower( $chunk );
 
@@ -285,10 +290,16 @@ function map_glossary_entries_to_translation_originals( $translation, $glossary
 		if ( is_array( $plural_split ) ) {
 			$plural_combined = '';
 
-			foreach ( $plural_split as $chunk ) {
+			foreach ( $plural_split as $k => $chunk ) {
 				// Create an escaped version for use later on.
 				$escaped_chunk = esc_translation( $chunk );
 
+
+				if ( $k > 0 && ( '<' === substr( $plural_split[ $k - 1 ], -1 ) || '"' === substr( $plural_split[ $k - 1 ], -1 ) || '</' === substr( $plural_split[ $k - 1 ], -2 ) ) ) {
+					$plural_combined .= $escaped_chunk;
+					continue;
+				}
+
 				// Create a lower case version to compare with the glossary terms.
 				$lower_chunk = strtolower( $chunk );
```

amieiro committed:
Merge branch 'develop' into tooltip-html-tag

amieiro committed:
Remove the regex added before

Props @akirk at https://github.com/GlotPress/GlotPress/pull/1679#issuecomment-1708103151

amieiro commented:
@akirk with the current approach we don't show the glossary suggestions into some elements like `title` or `alt` elements.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/80f394ff-3cd5-4a47-ad78-b49e726be743)

I think the best approach will be to remove these checks:

```
'"' === substr( $singular_split[ $k - 1 ], -1 ) ||
```

```
 '"' === substr( $plural_split[ $k - 1 ], -1 ) ||
``` 

![image](https://github.com/GlotPress/GlotPress/assets/1667814/5b512bd9-9e15-4e69-9ee2-0600d644daf6)

And then focus on special situations like `src` into the `img` tag, `href` into the `a` tag and another similar situations.

akirk commented:
We can change it to something like `('"' === substr( $singular_split[ $k - 1 ], -1 ) && ! ( 'alt="' === substr( $singular_split[ $k - 1 ], -5 ) || 'title="' === substr( $singular_split[ $k - 1 ], -7 ) ) )` etc. Best to refactor it into a new function.

amieiro committed:
Maintain the alt and title elements with the glossary helper

amieiro commented:
Now it maintains the glossary helper inside the HTML tags with the `alt` and `title` elements.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/e363e9f6-6586-4a3f-af2c-58f4a8f7274a)

amieiro committed:
Merge branch 'develop' into tooltip-html-tag

akirk commented on gp-templates/helper-functions.php:
```diff
@@ -514,3 +524,23 @@ function gp_translations_bulk_actions_toolbar( $bulk_action, $can_write, $transl
+ *
+ * @return bool
+ */
+function should_skip_element( int $key, array $split_element ): bool {
+	if ( $key > 0 &&
+		( '<' === substr( $split_element[ $key - 1 ], -1 ) ||
+			( '"' === substr( $split_element[ $key - 1 ], -1 ) && ! ( 'alt="' === substr( $split_element[ $key - 1 ], -5 ) || 'title="' === substr( $split_element[ $key - 1 ], -7 ) ) ) ||
+			'</' === substr( $split_element[ $key - 1 ], -2 ) ) ) {
+		return true;
+	}
+	return false;
```
I am a fan of keeping nesting levels low and ifs simple. You could refactor it to this:
```suggestion
function should_skip_element( int $key, array $split_element ): bool {
	if ( $key <= 0 ) {
		// We cannot skip the first element.
		return false;
	}
	$previous_element = $split_element[ $key - 1 ];
	$previous_character = substr( $previous_element, -1 );

	// Glossary word is an HTML tag because it is preceeded with a <, e.g. <strong>.
	if ( '<' === $previous_character ) {
		 return true;
	}

	// Glossary word is an HTML end tag because it is preceeded with a </, e.g. </strong>.
	if ( '</' === substr( $previous_element, -2 ) ) {
		 return true;
	}

	// If the preceeding char is a quotation mark, we might be in an HTML attribute.
	if ( '"' === $previous_character || "'" === $previous_character ) {
		// Don't skip if we're in an alt attribute.
		if ( 'alt=' === strtolower( substr( $previous_element, -5, 4 ) ) ) {
			return false;
		}

		// Don't skip if we're in a title attribute.
		if ( 'title=' === strtolower( substr( $previous_element, -7, 6 ) ) ) {
			return false;
		}

		// Skip if we're in any other attribute.
		return true;
	}

	// Don't skip in any other cases.
	return false;
```

Maybe we can also put this function inside the `map_glossary_entries_to_translation_originals` since we don't need it anywhere else and thus would not bloat the global namespace.

akirk replied:
Since we only need `$previous_element = $split_element[ $key - 1 ]` we could make that the function parameter as well but that would shift some complexity to the loop. We also don't have to go the `$key - 1` route but could do something like
```php
$previous_element = '';
foreach ( $singular_split as $chunk ) {
	if ( $this->should_skip_element( $previous_element ) ) {
		$previous_element = $chunk;
		continue;
	}
	// ...
	$previous_element = $chunk;
}
```

and
```php
function should_skip_element( string $previous_element ): bool {
	if ( ! $previous_element ) {
		// We cannot skip the first element.
		return false;
	}
	$previous_character = substr( $previous_element, -1 );
	// ...
}
```

akirk reviewed (commented):
I think making that function simpler would be good.

amieiro committed:
Update the should_skip_element method to resolve some bugs

amieiro committed:
Merge branch 'develop' into tooltip-html-tag

amieiro commented:
With the previous approach used in the ` should_skip_element` function, we detect some glossary word that should be skipped:

![image](https://github.com/GlotPress/GlotPress/assets/1667814/649d9fad-b591-4932-8330-6fb1a5152160)

With the new approach, we resolve these false positives:

![image](https://github.com/GlotPress/GlotPress/assets/1667814/d690d15d-734c-4783-9a42-7dbf47570335)

I prefer to put this code in a new function because we use this code twice in the `map_glossary_entries_to_translation_originals` function, and to avoid increasing the size of this function (currently, it has 193 lines).

akirk commented on gp-templates/helper-functions.php:
```diff
@@ -536,11 +536,78 @@ function gp_translations_bulk_actions_toolbar( $bulk_action, $can_write, $transl
+		}
+
+		// Check for the attribute elements inside the tag
+		if ( true === $insideTag ) {
```
Why not return early if we are not inside a tag?

akirk commented on gp-templates/helper-functions.php:
```diff
@@ -225,8 +225,8 @@ function map_glossary_entries_to_translation_originals( $translation, $glossary
 	$singular_split = preg_split( '/' . $terms_search . '/i', $translation->singular, 0, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE );
-
 	// Loop through each chunk of the split to find glossary terms.
+	// dd($translation->singular, $singular_split);
```
Stray dd

akirk committed:
Use a state machine

akirk reviewed (commented):
Sorry but I am still a bit unhappy about the complexity of the function. Check how it can be done with a state machine here: https://github.com/GlotPress/GlotPress/commit/29bc1c468020e03269e5ef89692540d16d71d69d

amieiro commented:
@akirk I review your commit https://github.com/GlotPress/GlotPress/commit/29bc1c468020e03269e5ef89692540d16d71d69d, and I think your approach is better, so, can you add this commit to this PR to close it?

akirk commented:
Could you cherry-pick the commit?
